lib/srv/desktop/windows_server.go

```diff
@@ -0,0 +1,330 @@
+		AnnouncePeriod:  apidefaults.ServerAnnounceTTL/2 + utils.RandomDuration(apidefaults.ServerAnnounceTTL/10),
+		CheckPeriod:     defaults.HeartbeatCheckPeriod,
+		ServerTTL:       apidefaults.ServerAnnounceTTL,
+		OnHeartbeat:     s.cfg.Heartbeat.OnHeartbeat,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	go func() {
+		if err := heartbeat.Run(); err != nil {
+			s.cfg.Log.WithError(err).Error("Heartbeat ended with error")
+		}
+	}()
+	return nil
+}
+
+// startStaticHostHeartbeats spawns heartbeat routines for all static hosts in
+// this service. We use heartbeats instead of registering once at startup to
+// support expiration.
+//
+// When a WindowsService with a list of static hosts disappears, those hosts
+// should eventually get cleaned up. But they should exist as long as the
+// service itself is running.
+func (s *WindowsService) startStaticHostHeartbeats() error {
+	for _, host := range s.cfg.Heartbeat.StaticHosts {
+		heartbeat, err := srv.NewHeartbeat(srv.HeartbeatConfig{
+			Context:         s.closeCtx,
+			Component:       teleport.ComponentWindowsDesktop,
+			Mode:            srv.HeartbeatModeWindowsDesktop,
+			Announcer:       s.cfg.AccessPoint,
+			GetServerInfo:   s.getHostHeartbeatInfo(host),
+			KeepAlivePeriod: apidefaults.ServerKeepAliveTTL,
+			AnnouncePeriod:  apidefaults.ServerAnnounceTTL/2 + utils.RandomDuration(apidefaults.ServerAnnounceTTL/10),
+			CheckPeriod:     defaults.HeartbeatCheckPeriod,
+			ServerTTL:       apidefaults.ServerAnnounceTTL,
+			OnHeartbeat:     s.cfg.Heartbeat.OnHeartbeat,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		go func() {
+			if err := heartbeat.Run(); err != nil {
+				s.cfg.Log.WithError(err).Error("Heartbeat ended with error")
+			}
+		}()
+	}
+	return nil
+}
+
+// Close instructs the server to stop accepting new connections and abort all
+// established ones. Close does not wait for the connections to be finished.
+func (s *WindowsService) Close() error {
+	s.close()
+	return nil
+}
+
+// Serve starts serving TLS connections fro plainLis. plainLis should be a TCP
+// listener and Serve will handle TLS internally.
+func (s *WindowsService) Serve(plainLis net.Listener) error {
+	lis := tls.NewListener(plainLis, s.cfg.TLS)
+	defer lis.Close()
+	for {
+		select {
+		case <-s.closeCtx.Done():
+			return trace.Wrap(s.closeCtx.Err())
+		default:
+		}
+
+		con, err := lis.Accept()
+		if err != nil {
+			if utils.IsOKNetworkError(err) || trace.IsConnectionProblem(err) {
+				return nil
+			}
+			return trace.Wrap(err)
+		}
+
+		go s.handleConnection(con)
+	}
+}
+
+func (s *WindowsService) handleConnection(con net.Conn) {
+	defer con.Close()
+
+	// Check connection limits.
+	remoteAddr, _, err := net.SplitHostPort(con.RemoteAddr().String())
+	if err != nil {
+		s.cfg.Log.WithError(err).Errorf("Could not parse client IP from %q", con.RemoteAddr().String())
+		return
+	}
+	if err := s.cfg.ConnLimiter.AcquireConnection(remoteAddr); err != nil {
```

Comment:
It might make sense for these limiters to export a metric for how many connections they reject. Would be a good stat to have when monitoring a large environment.

Reply:
Agreed. Filed an issue at https://github.com/gravitational/teleport/issues/7953 to collect those centrally.